backend/fms_core/services/sample.py

```diff
@@ -443,10 +443,76 @@ def remove_qc_flags(sample):
         sample.quality_flag = None
         sample.save()
     except Error as e:
-        errors.appends(';'.join(e.messages))
+        errors.append(';'.join(e.messages))
 
     return sample, errors, warnings
 
 
+def add_sample_properties(sample, properties):
+    errors = []
+    warnings = []
+
+    if sample and properties:
+        try:
+            for property_name in properties.keys():
+                property_value = properties[property_name]
+                # Check if sample has already the property
+                if SampleProperty.objects.filter(name=property_name, sample_id=sample.id).exists():
+                    errors.append(f"[Sample {sample.name}] already has property [{property_name}] with value [{property_value}].")
+                else:
+                    SampleProperty.objects.create(name=property_name, value=property_value, sample=sample)
+        except ValidationError as e:
+            errors.append(';'.join(e))
+    else:
+        errors.append('Sample and properties are required')
+
+    return properties, errors, warnings
+
+
+def update_sample_properties(sample, properties):
+    errors = []
+    warnings = []
+
+    if sample and properties:
+        try:
+            for property_name in properties.keys():
+                property_value = properties[property_name]
+                # Check if sample has already the property
+                if SampleProperty.objects.filter(name=property_name, sample_id=sample.id).exists():
+                    property = SampleProperty.objects.get(name=property_name, value=property_value, sample=sample)
+                    property.value = property_value
+                    property.save()
+                else:
+                    errors.append(f"[Sample {sample.name}] does not have property [{property_name}].")
+                SampleProperty.objects.create(name=property_name, value=property_value, sample=sample)
+        except ValidationError as e:
+            errors.append(';'.join(e))
+    else:
+        errors.append('Sample and properties are required')
+
+    return properties, errors, warnings
+
+
+def remove_sample_properties(sample, properties):
+    num_objects_deleted = 0
+    errors = []
+    warnings = []
+
+    if sample and properties:
+        try:
+            for property_name in properties.keys():
+                property_value = properties[property_name]
+                num_objects_deleted, _ = SampleProperty.objects.filter(name=property_name, value=property_value, sample=sample).delete()
```

Comment:
I don't think this will call the model delete() which is the one we need to call to set the delete flag in the trackingfields so we record the current value in reversion. Since this can apply delete to multiple delete this will likely call the sampleProperty manager delete() instead (which we haven't overloaded). You can double check if you want to confirm and not go just by my word. I think calling first() after filter would remedy this.